substrate/client/transaction-pool/src/common/api.rs

```diff
@@ -39,38 +37,85 @@ use sp_runtime::{
 	transaction_validity::{TransactionSource, TransactionValidity},
 };
 use sp_transaction_pool::runtime_api::TaggedTransactionQueue;
+use std::{
+	marker::PhantomData,
+	pin::Pin,
+	sync::Arc,
+	time::{Duration, Instant},
+};
+use tokio::sync::{mpsc, oneshot, Mutex};
 
 use super::{
 	error::{self, Error},
 	metrics::{ApiMetrics, ApiMetricsExt},
 };
 use crate::graph;
-use tracing::{trace, warn};
+use tracing::{instrument, trace, warn, Level};
 
 /// The transaction pool logic for full client.
 pub struct FullChainApi<Client, Block> {
 	client: Arc<Client>,
 	_marker: PhantomData<Block>,
 	metrics: Option<Arc<ApiMetrics>>,
-	validation_pool: mpsc::Sender<Pin<Box<dyn Future<Output = ()> + Send>>>,
+	validation_pool_normal: mpsc::Sender<Pin<Box<dyn Future<Output = ()> + Send>>>,
+	validation_pool_maintained: mpsc::Sender<Pin<Box<dyn Future<Output = ()> + Send>>>,
+	validate_transaction_normal_stats: DurationSlidingStats,
+	validate_transaction_maintained_stats: DurationSlidingStats,
 }
 
 /// Spawn a validation task that will be used by the transaction pool to validate transactions.
 fn spawn_validation_pool_task(
 	name: &'static str,
-	receiver: Arc<Mutex<mpsc::Receiver<Pin<Box<dyn Future<Output = ()> + Send>>>>>,
+	receiver_normal: Arc<Mutex<mpsc::Receiver<Pin<Box<dyn Future<Output = ()> + Send>>>>>,
+	receiver_maintained: Arc<Mutex<mpsc::Receiver<Pin<Box<dyn Future<Output = ()> + Send>>>>>,
 	spawner: &impl SpawnEssentialNamed,
+	stats: DurationSlidingStats,
+	blocking_stats: DurationSlidingStats,
 ) {
 	spawner.spawn_essential_blocking(
 		name,
 		Some("transaction-pool"),
 		async move {
 			loop {
-				let task = receiver.lock().await.next().await;
-				match task {
-					None => return,
-					Some(task) => task.await,
-				}
+				let start = Instant::now();
+
+				let task = {
+					let receiver_maintained = receiver_maintained.clone();
+					let receiver_normal = receiver_normal.clone();
+					tokio::select! {
```

Comment:
When both `receiver_maintained` and `receiver_normal` are ready, tokio will choose a task at random. Is this your intention? I think you might want to use `biased` here to establish some priority.

Reply:
This exmple shall explain your concern:
Imagine the queue, (in the previous imeplementation): `SSSSSSSSSSSSSSSSSMM`. The `M` transactions will be validated after validation ALL `S` transactions.

With two queues, each 50%, the validation will of `M`  transaction validation will be perorfmed at the same pace as `S`. They will not get blocked by numerous `S` transactions. This should be enough for now.